src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+                            ? description
+                            : string.Empty))
+                    : null;
+
+            session.Properties.AddProperty(MustSetSelection, suggestionItemOptions != null);
+
+            return new EditorCompletion.CompletionContext(
+                items,
+                suggestionItemOptions,
+                suggestionItemOptions == null ? EditorCompletion.InitialSelectionHint.RegularSelection : EditorCompletion.InitialSelectionHint.SoftSelection);
+        }
+
+        public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, EditorCompletion.CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(RoslynItem, out var roslynItem) ||
+                !item.Properties.TryGetProperty<ITextBuffer>(TriggerBuffer, out var triggerBuffer))
+            {
+                return null;
+            }
+
+            var document = triggerBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return null;
+            }
+
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            var description = await service.GetProvider(roslynItem).GetDescriptionAsync(document, roslynItem, cancellationToken).ConfigureAwait(false);
+
+            return new ClassifiedTextElement(description.TaggedParts.Select(p => new ClassifiedTextRun(p.Tag.ToClassificationTypeName(), p.Text)));
+        }
+
+        private static RoslynTrigger GetRoslynTrigger(EditorCompletion.CompletionTrigger trigger, SnapshotPoint triggerLocation)
+        {
+            switch (trigger.Reason)
+            {
+                case EditorCompletion.CompletionTriggerReason.Invoke:
+                case EditorCompletion.CompletionTriggerReason.InvokeAndCommitIfUnique:
+                    return RoslynTrigger.Invoke;
+                case EditorCompletion.CompletionTriggerReason.Insertion:
+                    return RoslynTrigger.CreateInsertionTrigger(trigger.Character);
+                case EditorCompletion.CompletionTriggerReason.Deletion:
+                    return RoslynTrigger.CreateDeletionTrigger(trigger.Character);
+                case EditorCompletion.CompletionTriggerReason.Backspace:
+                    var snapshotBeforeEdit = trigger.ViewSnapshotBeforeTrigger;
+                    char characterRemoved;
+                    if (triggerLocation.Position >= 0 && triggerLocation.Position < snapshotBeforeEdit.Length)
+                    {
+                        // If multiple characters were removed (selection), this finds the first character. 
+                        // Maybe it should be re-considered to find the last removed character.
+                        characterRemoved = snapshotBeforeEdit[triggerLocation.Position];
```

Comment:
i think we need to be far more explicit about finding just an exact backspace.  if multiple things are backspaced, should we even be bringing up completion?  what's our current behavior?

Reply:
Thank you! Added a test to compare both behaviors (TestBackspaceWithMultipleCharactersSelected) and simplified the comment.

---
In reply to: [222083202](https://github.com/dotnet/roslyn/pull/29016#discussion_r222083202) [](ancestors = 222083202)